pkg/services/notary/notary.go

```diff
@@ -0,0 +1,376 @@
+		if err := n.finalize(r.main); err != nil {
+			n.Config.Log.Error("failed to finalize main transaction", zap.Error(err))
+		} else {
+			r.isSent = true
+		}
+	}
+}
+
+// OnRequestRemoval is a callback which is called after fallback transaction is removed
+// from the notary payload pool due to expiration, main tx appliance or any other reason.
+func (n *Notary) OnRequestRemoval(pld *payload.P2PNotaryRequest) {
+	if n.getAccount() == nil {
+		return
+	}
+
+	n.reqMtx.Lock()
+	defer n.reqMtx.Unlock()
+	r, ok := n.requests[pld.MainTransaction.Hash()]
+	if !ok {
+		return
+	}
+	for i, fb := range r.fallbacks {
+		if fb.Hash().Equals(pld.FallbackTransaction.Hash()) {
+			r.fallbacks = append(r.fallbacks[:i], r.fallbacks[i+1:]...)
+			break
+		}
+	}
+	if len(r.fallbacks) == 0 {
+		delete(n.requests, r.main.Hash())
+	}
+}
+
+// PostPersist is a callback which is called after new block is persisted.
+func (n *Notary) PostPersist(bc blockchainer.Blockchainer, pool *mempool.Pool, b *block.Block) {
+	if n.getAccount() == nil {
+		return
+	}
+
+	n.reqMtx.Lock()
+	defer n.reqMtx.Unlock()
+	for h, r := range n.requests {
+		if !r.isSent && r.typ != Unknown && r.nSigs == r.nSigsCollected && r.minNotValidBefore > bc.BlockHeight() {
+			if err := n.finalize(r.main); err != nil {
+				n.Config.Log.Error("failed to finalize main transaction", zap.Error(err))
+			} else {
+				r.isSent = true
+			}
+			continue
+		}
+		if r.minNotValidBefore <= bc.BlockHeight() { // then at least one of the fallbacks can already be sent.
+			newFallbacks := r.fallbacks[:0]
+			for _, fb := range r.fallbacks {
+				if nvb := fb.GetAttributes(transaction.NotValidBeforeT)[0].Value.(*transaction.NotValidBefore).Height; nvb <= bc.BlockHeight() {
+					if err := n.finalize(fb); err != nil {
+						newFallbacks = append(newFallbacks, fb) // wait for the next block to resend them
+					}
+				} else {
+					newFallbacks = append(newFallbacks, fb)
+				}
+			}
+			if len(newFallbacks) == 0 {
+				delete(n.requests, h)
+			} else {
+				r.fallbacks = newFallbacks
+			}
+		}
+	}
+}
+
+// finalize adds missing Notary witnesses to the transaction (main or fallback) and pushes it to the network.
+func (n *Notary) finalize(tx *transaction.Transaction) error {
+	acc := n.getAccount()
+	if acc == nil {
+		panic(errors.New("no available Notary account"))
+	}
+	notaryWitness := transaction.Witness{
+		InvocationScript:   append([]byte{byte(opcode.PUSHDATA1), 64}, acc.PrivateKey().Sign(tx.GetSignedPart())...),
+		VerificationScript: []byte{},
+	}
+	for i, signer := range tx.Signers {
+		if signer.Account == n.Config.Chain.GetNotaryContractScriptHash() {
+			tx.Scripts[i] = notaryWitness
+			break
+		}
+	}
+	return n.getOnTransaction()(tx)
+}
+
+// SetOnTransaction sets callback method which should be called after transaction is completed.
+func (n *Notary) SetOnTransaction(f func(tx *transaction.Transaction) error) {
```

Comment:
Do we ever change `onTransaction` in runtime? Maybe it can be a parameter for initialization?

Reply:
No, we don't. Fixed.